README.md

```diff
@@ -376,6 +376,125 @@ Binaries produced by the Swift build process must be generated for the correct p
 docker run --rm -it -v $(pwd):/swift-package openwhisk/action-swift-v3.1.1 bash -e -c "cd /swift-package && swift build -v -c release"
 ```
 
+## Writing Functions - Java
+
+Here's an `src/main/java/HelloWorld.java` file containing an example handler function.
+
+```java
+import com.google.gson.JsonObject;
+
+public class HelloWorld {
+
+  public static JsonObject main(JsonObject args) throws Exception {
+
+    final String name = args.getAsJsonPrimitive("name").getAsString();
+
+    final JsonObject response = new JsonObject();
+    response.addProperty("greeting", "Hello " + name + "!");
+
+    return response;
+  }
+}
+```
+
+Here is a simple `pom.xml` file that will allow you to use Maven to build it. You will notice that `gson` is excluded from the uberjar. That is because OpenWhisk already provides this dependency.
+
+```xml
+<project>
+ <modelVersion>4.0.0</modelVersion>
+ <groupId>hello</groupId>
+ <artifactId>hello-world</artifactId>
+ <version>1.0</version>
+
+ <dependencies>
+  <dependency>
+    <groupId>com.google.code.gson</groupId>
+    <artifactId>gson</artifactId>
+    <version>2.8.2</version>
+  </dependency>
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-shade-plugin</artifactId>
+        <version>3.1.0</version>
+        <executions>
+          <execution>
+            <phase>package</phase>
+            <goals>
+              <goal>shade</goal>
+            </goals>
+            <configuration>
+              <minimizeJar>true</minimizeJar>
+              <artifactSet>
+                <excludes>
+                  <exclude>com.google.code.gson:gson</exclude>
+                </excludes>
+              </artifactSet>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+ </project>
+```
+
+In the `serverless.yaml` file (see below), the `handler` property is the uberjar produced by calling `mvn clean package`. To specify the name of the class containing the `main` function, add an `annotation` called `main_class` (this should be the fully qualified class name). If you do not provide this `annotation`, it will look for a class in the default package called `Main`.
+
+```yaml
+service: my-java-service
+provider:
+  name: openwhisk
+  runtime: java
+functions:
+  hello:
+    handler: target/hello-world-1.0.jar
```

Comment:
I'd prefer to provide both the JAR file and the main class in the `handler` parameter. This keeps it consistent with the other runtimes which use a `file.method` style. For this runtime, we would adopt the following:
```
    handler: target/hello-world-1.0.HelloWorld
```

This should be a small change to your PR - I think if you back out the convertHandlerToPath & calculateFunctionMain fns it should just work?

Reply:
That is going to get a little gnarly if they follow the standard practice of NOT using the default package and actually place their class into a package. For example, if they put the class in a package called `com.company.app` then the handler will look like this:

```
handler: target/hello-world-1.0.com.company.app.HelloWorld
```

That is visually a bit hard to parse (and I THINK breaks the code you were referring to). I suppose you could do something like this:

```
handler: target/hello-world-1.0.jar:com.company.app.HelloWorld
```

The reason I used annotation in this pull request was I saw that was how you marked an action as a web action... So I was thinking annotations could be used to provide further refinements of how actions needed to be deployed. That being said, I understand what you mean by consistency, so definitely open to changing that part.

Reply:
I can see that might get a bit long. `handler: target/hello-world-1.0.jar:com.company.app.HelloWorld` is definitely a bit better style. 

[Annotations](https://github.com/apache/incubator-openwhisk/blob/master/docs/annotations.md) are actually passed through to the platform when creating the actions. They aren't used for meta-data inside the framework if you see what I mean? 

I agree it does get a bit long but think the second style (with `:`) is the best compromise between consistency and readability.